text/1002-optionality-qualifier.rst

```diff
@@ -0,0 +1,200 @@
+2. client code written to insert data to the database or update data in
+   it might become invalid if it depended on the optionality.
+
+The former issue may be addressed with a data migration prior to the
+schema migration.  The latter problem is out of scope for EdgeDB but
+can be handled by migrating client code *first*.  If the clients aren't
+under the database administrator's control, a reasonable choice in this
+case is to abort the attempt to make a pointer required.
+
+Additionally, optionality by default in pointers is inconsistent with
+non-optionality in function arguments.
+
+
+Rejected ideas
+==============
+
+Assume REQUIRED qualifiers when no explicit qualifier is provided
+-----------------------------------------------------------------
+
+An extreme form of optionality everywhere is equivalent to the database
+not having a true schema. Unless it's necessary to make a pointer optional,
+forcing pointers to be required ensures that whenever data *can* be specified,
+it won't be forgotten at INSERT or UPDATE time.
+
+Required pointers simplify type information. There's two particular problems
+on the client side with unexpected optionality:
+
+* lack of handling for that special case in user code, and
+
+* surprising effects of EdgeQL queries that omit objects with unexpected
+  empty sets in pointers used in the query.
+
+However, since many pointers in real-world schemas will be declared as OPTIONAL
+regardless of the default so the downsides of optionality wouldn't be
+solved by using REQUIRED as the default qualifier.
+
+Assuming that all pointers are required unless otherwise specified makes it
+easier to migrate into optionality later. Client code that was forced to
+provide values at INSERT time before will continue working and there are no
+incompatible empty sets in the database prior to the migration.  However,
+see "Downsides" below.
+
+Downsides
+~~~~~~~~~
+
+The main disadvantage at time of writing is that the current behavior as of
+EdgeDB v1.0a3 is the opposite: pointers without an explicit qualifier are
+assumed optional. This poses a `Migration challenge`_.
+
+Required pointers have to be always specified at INSERT time, making
+queries more verbose to write and data more costly to send over the wire.
+This can be dealt with by providing default values with SET DEFAULT which
+is a form of optionality with an explicit fallback value which doesn't fall
+outside of the declared type.
+
+Required pointers by default make it more likely that users would have to
+make required-to-optional pointer migrations in the future.  Those hide
+a specific gotcha: pre-existing queries that were previously guaranteed
+never to encounter empty sets now will.  Most of the time this will be
+harmless but some queries that are analytical in nature will now fail to
+return objects with empty pointers, which might lead to bugs in user code.
+
+Thrift documentation explains that "required is forever". What they mean
```

Comment:
Maybe move this entire paragraph into the "A database is not a peer-to-peer RPC platform" section?